core/arch/arm/kernel/generic_boot.c

```diff
@@ -551,6 +554,44 @@ static int add_optee_res_mem_dt_node(void *fdt)
 	return 0;
 }
 
+static void read_optee_secure_device_node(void *fdt)
+{
+	const uint32_t *p;
+	int offs;
+	int len;
+#if (TRACE_LEVEL >= TRACE_INFO)
+	char s[33];
+	int n = 0;
+	int m;
+#endif
+
+	offs = fdt_path_offset(fdt, "/firmware/optee");
+	if (offs < 0)
+		return;
+
+	p = fdt_getprop(fdt, offs, "secure-device-id", &len);
+	if (!p || !len)
+		return;
+
+	if (len > (int)sizeof(secure_device_id))
+		len = sizeof(secure_device_id);
+
+	memcpy(secure_device_id, p, len);
+	secure_device_id_len = len;
+
+#if (TRACE_LEVEL >= TRACE_INFO)
+	if (len > 16)
+		len = 16;
+
+	for (m = 0; m < len; m++)
+		n += snprintf(&s[n], sizeof(s) - n - 1, "%02X",
+			      secure_device_id[m]);
+
+	IMSG("secure-device-id: len %d: %s\n",
+	     (int)secure_device_id_len, s);
+#endif
```

Comment:
Inside OP-TEE there is `%pUl` that you can use with `snprintf()` and `DMSG()` and friends (this syntax was taken from the Linux kernel). When I added this format, I decided not to make it available to TAs, to avoid breaking the standard. Now I'm not sure it was a good choice...